### PR TITLE
backupccl: Fix MR enum in non-MR temp sys DB

### LINKED ISSUE #131285: ERROR: internal error: found MULTIREGION_ENUM on non-multi-region database crdb_temp_system
**Describe the problem**

A backup taken on a 24.1 Serverless cluster from crl-prod-j77 (AWS, us-east-1) was then attempted to be restored into a 24.2 OnPrem cluster (with a single locality of `region=us-east-1`).  During restore, this stack trace was produced:

```
root@localhost:26257/system> RESTORE FROM LATEST IN  .... s3 details elided...

ERROR: internal error: found MULTIREGION_ENUM on non-multi-region database crdb_temp_system
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1229: func1()
github.com/cockroachdb/cockroach/pkg/sql/internal.go:1930: func3()
github.com/cockroachdb/cockroach/pkg/sql/internal.go:2027: func4()
github.com/cockroachdb/cockroach/pkg/kv/txn.go:1049: exec()
github.com/cockroachdb/cockroach/pkg/kv/db.go:1103: runTxn()
github.com/cockroachdb/cockroach/pkg/kv/db.go:1066: TxnWithAdmissionControl()
github.com/cockroachdb/cockroach/pkg/kv/db.go:1041: Txn()
github.com/cockroachdb/cockroach/pkg/sql/internal.go:2014: txn()
github.com/cockroachdb/cockroach/pkg/sql/internal.go:1928: DescsTxn()
github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1190: createImportingDescriptors()
github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1782: doResume()
github.com/cockroachdb/cockroach/pkg/ccl/backupccl/restore_job.go:1720: Resume()
github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1632: func2()
github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1633: stepThroughStateMachine()
github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:451: runJob()
github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:831: func2()
github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485: func2()
src/runtime/asm_amd64.s:1695: goexit()

HINT: You have encountered an unexpected error.
```

**To Reproduce**

* Create a single node 24.2.2 cluster, with the node having a locality of `--locality=region=aws-us-east-1`.
* RESTORE FROM ... (contact me separately for the AWS keys and details).

**Expected behavior**
The restore should work as normal.

**Environment:**
 - CockroachDB version: 24.1 backup, 24.2 restore.

**Additional context**
Restore could not be taken.

* There's the possibility that the backup was taken while the host cluster was upgrading from 24.1 to 24.2.  Although an earlier error message shows that the backup was taken on a 24.1 cluster.

Jira issue: CRDB-42475

### PR DESCRIPTION
**Pull request by dt**
The temp db doesn't need to be MR just for us to copy stuff out of it into the real one.

Release note (bug fix): Fix a bug that prevented restoring cluster backup taken in an MR cluster that had configured a the system database with a region configuration into a non-multi-region cluster.
Epic: CRDB-42475.

### CODE REVIEW

**Review by fqazi — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134604)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @kev-cao)

___
*[`pkg/ccl/backupccl/restore_job.go` line 1254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134604#-OBB5dTK2H4LW79ULYP7:-OBB5dTK2H4LW79ULYP8:b-hny093) ([raw file](https://github.com/cockroachdb/cockroach/blob/f041746f1eb0b756af6414089dbfff55771b9834/pkg/ccl/backupccl/restore_job.go#L1254)):*
> ```Go
> 						t.TypeDesc().RegionConfig = nil
> 						// TODO(foundations): should these be rewritten instead of blank? Does it matter since we drop the whole DB before the job exits?
> 						t.TypeDesc().ReferencingDescriptorIDs = nil
> ```

I guess this makes validation happy? How is the temporary database cleaned up?

**Comment by dt** [comment]
We drop the whole temp DB later in the job.

File: pkg/ccl/backupccl/restore_job.go [hunk 1]
```diff
@@ -1179,6 +1179,7 @@ func createImportingDescriptors(
 		for _, desc := range mutableTables {
 			if desc.GetParentID() == tempSystemDBID {
 				desc.SetPublic()
+				desc.LocalityConfig = nil
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_job.go:1182]: why are you niling this out for all tables?

**Review by msbutler — APPROVED** [review]
i'm a bit out of my league here, but happy to support the cause.

**Comment by dt** [comment]
TL;DR is that for _cluster_ restores we make a temp database, restore backed up _system_ tables into it so we can read them to manipulate the real system tables, then drop that database and all the tables we restored into it once we're done. 

This isn't new / is working fine.

At some point (@jeffswenson can probably chime in here), in serverless they started making the system database multi-region, and some of the tables in it multi-region as well. Our temp database is _not_ multiregion and never has been, but this  trips up the catalog when we try to stick a backed up MR system table in a non-MR database. We don't need these tables to be MR -- we're gonna access them once, from one function, then drop them. So now we de-MR them to put them in our non-MR tmp db.

File: pkg/ccl/backupccl/restore_job.go (hunk 1 above)
Comment (jeffswenson) [reply pkg/ccl/backupccl/restore_job.go:1182]: I think this is sound since we aren't trying to write any sessions, leases, or sql_instances. So the fact that the table is or is not mr shouldn't matter to any of the data we are restoring.

**Review by jeffswenson — APPROVED** [review]
LGTM

**Comment by dt** [comment]
TFTR!

bors r+

**Comment by dt** [comment]
blathers backport 24.3

**Comment by dt** [comment]
blathers backport 24.2

**Comment by dt** [comment]
blathers backport 24.1
